### PR TITLE
Introduce RPC based prover to kevm

### PR DESCRIPTION
**Pull request by ehildenb**
~Blocked on: https://github.com/runtimeverification/evm-semantics/pull/1527~
~Blocked on: https://github.com/runtimeverification/pyk/pull/173~
~Blocked on: https://github.com/runtimeverification/evm-semantics/pull/1544~
~Blocked on: https://github.com/runtimeverification/evm-semantics/pull/1560~

This PR makes several steps to making RPC based proving possible with KEVM:

- Names the various `CALL*` rules so that they can be broken on.
- Makes use of new upstreamed RPC based prover in all pyk-based prover (including `kevm prove ... --pyk` and `kevm foundry-prove ...`). These come with several options for controlling execution/KCFG construction:
  - `--max-iterations`: maximum number of times to expand the KCFG created between initial and target states in an attempt to show all-path reachability.
  - `--max-depth`: maximum K steps to take for each leg of symbolic execution when expanding the KCFG from the initial state to the target state.
  - `--[no-]implication-every-block`: check implication every time that a state is returned from symbolic execution (or not).
  - `--break-on-calls`: save the pre-state every time there is a contract call, from right before the call.
  - `--break-every-step`: save a state every opcode that is executed.
- Adds command `kevm foundry-remove-node ...` which allows remove nodes (and successors) from a KCFG so you can restart from a prior state with lemmas added, for example.
- Adds command `kevm foundry-simplify-node ... [--replace]` which allows calling the simplification routine on a given node and printing the results, and optionally also replacing the given node in eth KCFG.
- Adds command `kevm view-kcfg --definition path/to/kompiled path/to/kcfg-file.json`, for viewing KCFGS produced with `kevm prove .... --pyk --save-directory path/to`.
- Enables the `--pyk` option for as many haskell proofs as possible on CI, lists the ones failing with pyk in `tests/failing-symbolic.pyk`.
- Increases the parallelism for both Haskell and Foundry proofs on CI due to lower overhead of RPC based proving.
- Disables two foundry tests that actually were being vacuously proved by reaching `#Bottom`.
- Corrects the way that logging levels are handled.
- Passes in the `KEVM.short_info` as a pretty-printer for the KCFG behavior view, and for viewing KCFGs in `foundry-kcfg-show`.

Note that CI time is reduced to ~35 minutes for Haskell backend proofs, and just over an hour for Foundry proofs (down from 3.5-4 hours!).

### CODE REVIEW

File: .github/workflows/test-pr.yml [hunk 1]
```diff
@@ -109,7 +109,7 @@ jobs:
       - name: 'Build Foundry'
         run: docker exec -u user kevm-ci-foundry-${GITHUB_SHA} /bin/bash -c 'make build-foundry -j2'
       - name: 'Test Foundry'
-        run: docker exec -u user kevm-ci-foundry-${GITHUB_SHA} /bin/bash -c 'make test-foundry -j2 FOUNDRY_PAR=4 KPROVE_OPTS=--retry'
```
Comment (anvacaru) [inline .github/workflows/test-pr.yml:112]: I'd suggest removing the `retry` feature since it was only added for CI.

File: bin/kevm [hunk 2]
```diff
@@ -391,6 +391,7 @@ kevm_port='8545'
 kevm_host='127.0.0.1'
 [[ ! "${run_command}" == prove     ]] || backend=haskell
 [[ ! "${run_command}" == solc-to-k ]] || backend=haskell
+[[ ! "${run_command}" == view-kcfg ]] || backend=haskell
```
Comment (anvacaru) [inline bin/kevm:394]: Should `view-kcfg` be used with `--backend foundry` when it is used with a foundry proof?

File: kevm-pyk/src/kevm_pyk/kevm.py [hunk 3]
```diff
@@ -217,7 +235,70 @@ def add_invariant(cterm: CTerm) -> CTerm:
...
+            ):
+                term = remove_constraints_for(['GAS_CELL'], term)
+                if type(gas_cell) is KApply and gas_cell.label.name == 'infGas':
+                    term = set_cell(term, 'GAS_CELL', KEVM.inf_gas(KVariable('GAS_CELL')))
+                else:
+                    term = set_cell(term, 'GAS_CELL', KVariable('GAS_CELL'))
+        # memoryused_cell = get_cell(term, 'MEMORYUSED_CELL')
+        # if type(memoryused_cell) is not KVariable or count_vars(term)[memoryused_cell.name] != 1:
+        #     term = remove_constraints_for(['MEMORYUSED_CELL'], term)
+        #     term = set_cell(term, 'MEMORYUSED_CELL', KVariable('MEMORYUSED_CELL'))
```
Comment (anvacaru) [inline kevm-pyk/src/kevm_pyk/kevm.py:287]: Delete commented code maybe?

```suggestion
```

**Review by anvacaru — APPROVED** [review]
Approved with a few comments.

File: kevm-pyk/src/kevm_pyk/kevm.py [hunk 4]
```diff
@@ -217,7 +235,70 @@ def add_invariant(cterm: CTerm) -> CTerm:
...
+        wordstack_cell = get_cell(term, 'WORDSTACK_CELL')
+        KApply('.WordStack_EVM-TYPES_WordStack')
```
Comment (tothtamas28) [inline kevm-pyk/src/kevm_pyk/kevm.py:289]: Probably `autoflake` got rid of an unused variable here.

File: kevm-pyk/src/kevm_pyk/utils.py [hunk 5]
```diff
@@ -1,26 +1,113 @@
...
+        else:
+            _LOGGER.error(f'Proof failed: {_cfgid}')
+            return False
+
+    with ProcessPool(ncpus=workers) as process_pool:
+        _proof_problems = [(_id, _cfg, _i) for _i, (_id, _cfg) in enumerate(proof_problems.items())]
+        results = process_pool.map(_call_rpc, _proof_problems)
+
+    failed = 0
+    for cid, succeeded in zip(proof_problems.keys(), results, strict=True):
```
Comment (tothtamas28) [inline kevm-pyk/src/kevm_pyk/utils.py:99]:
```suggestion
    for cid, succeeded in zip(proof_problems, results, strict=True):
```

File: kevm-pyk/src/kevm_pyk/utils.py [hunk 6]
```diff
@@ -1,26 +1,113 @@
...
+        if succeeded:
+            print(f'PASSED: {cid}')
+        else:
+            print(f'FAILED: {cid}')
```
Comment (tothtamas28) [inline kevm-pyk/src/kevm_pyk/utils.py:103]: Consider pulling this side effect to the caller.

**Review by tothtamas28 — APPROVED** [review]

File: .github/workflows/test-pr.yml (hunk 1 above)
Comment (ehildenb) [reply .github/workflows/test-pr.yml:112]: It has been removed entirely here.

File: bin/kevm (hunk 2 above)
Comment (ehildenb) [reply bin/kevm:394]: No, we by default want this to work with any proof. No good way to detect it's a foundry proof. `foundry-view-kcfg` is the command to use for those ones, which does it's own argument processing.

File: .github/workflows/test-pr.yml (hunk 1 above)
Comment (anvacaru) [reply .github/workflows/test-pr.yml:112]: Yes, it was. My bad.

File: kevm-pyk/src/kevm_pyk/kevm.py (hunk 3 above)
Comment (ehildenb) [reply kevm-pyk/src/kevm_pyk/kevm.py:287]: Done.

File: kevm-pyk/src/kevm_pyk/kevm.py (hunk 4 above)
Comment (ehildenb) [reply kevm-pyk/src/kevm_pyk/kevm.py:289]: Removed.

File: kevm-pyk/src/kevm_pyk/utils.py (hunk 5 above)
Comment (ehildenb) [reply kevm-pyk/src/kevm_pyk/utils.py:99]: Done.

File: kevm-pyk/src/kevm_pyk/utils.py (hunk 6 above)
Comment (ehildenb) [reply kevm-pyk/src/kevm_pyk/utils.py]: Good idea, done!

File: kevm-pyk/src/kevm_pyk/utils.py (hunk 5 above)
Comment (ehildenb) [reply kevm-pyk/src/kevm_pyk/utils.py]: Done.

File: kevm-pyk/src/kevm_pyk/kevm.py (hunk 4 above)
Comment (ehildenb) [reply kevm-pyk/src/kevm_pyk/kevm.py]: Fixed.

File: kevm-pyk/src/kevm_pyk/kevm.py (hunk 3 above)
Comment (ehildenb) [reply kevm-pyk/src/kevm_pyk/kevm.py]: Removed.

**Comment by ehildenb** [comment]
After @PetarMax discovered a bug in implication checking, I'd rather wait for the fix to make it to here to make sure the proofs I've marked as passing on this PR are actually passing.

**Pull request merged**